### ISSUE TITLE
[BUG]: span-overflow auto-tiling can't tile large matmuls (weight producer→matmul consumer not synchronized; disabled by default)

### ISSUE THREAD

**Issue by mudhakar** [torch-spyre/torch-spyre#3217]
### What happened?

The automatic span-overflow coarse-tiler (`span_overflow_hint_analysis.py` / `coarse_tile.span_overflow_groups`) is the mechanism intended to tile oversized matmuls so their per-core weight span stays within the 256 MB EAR limit (the fix path for #1918). Two problems block it for `F.linear`-shaped matmuls:

1. **It is disabled by default.** `config.ignore_span_overflow_hints` defaults to `"1"` (opt-in via `SPYRE_INDUCTOR_IGNORE_SPAN_OVERFLOW_HINTS=0`), so `span_overflow_groups` is never called and the op falls through to work division, which aborts with an EAR overflow (`dxp_standalone` SIGABRT) whenever the weight's outer stick dimension isn't work-division-splittable.

2. **When enabled, it cannot tile the matmul**, because the weight reaches the matmul through a Pointwise producer (restickify/layout op). The planner tiles the producer, then refuses to tile the matmul consumer:

   > `Unsupported: Cannot auto-tile buf0: it reads already auto-tiled producer(s) ['buf1']. Automatic span-overflow grouping currently only synchronizes compatible contiguous pointwise ops, so tiling this producer and consumer independently can produce unsynchronized loop nests.`

### Evidence — vocab sweep, `F.linear(x[1,4096], W[V,4096])`, `SENCORES=32`

Weight device layout `[V/64, 4096, 64]`; per-core weight span must be ≤ 256 MB.

| V | sticks | factorization | default (feature OFF) | feature ON (`...IGNORE_SPAN_OVERFLOW_HINTS=0`) |
|---|---|---|---|---|
| 49152 | 768 = 2⁸·3 | composite | **PASS** (work division splits d0÷2) | **FAIL** — `Cannot auto-tile buf0: it reads already auto-tiled producer(s) ['buf1']` |
| 49216 | 769 | **prime** | FAIL (work div can't split; EAR overflow) | FAIL — `no combined split among host dims [0] makes all spans fit` (buf1 `legal=[1]`; 769 has no stick-aligned split ≤ `_MAX_AUTO_TILE_SPLIT_COUNT=64`) |
| 87616 | 1369 = 37² | composite | FAIL (37 > 32 cores → work div can't split) | FAIL — same producer/consumer error, **even though the planner found a valid split** (`legal=[1, 37]`, 37-stick tiles = 19 MB) |

`V=87616` is the key case: work division cannot help (smallest divisor 37 > 32 cores) and the span-overflow planner **does** find a legal tile (`split=37`), but the compile still fails purely on the producer→consumer synchronization limitation — not on any span/cap constraint. So enabling the feature also **regresses** `V=49152`, which passes today via work division.

DEBUG trace (feature on, `V=87616`):
```
[span-overflow divisors] op=buf1 host_dim=0 required=3 legal=[1, 37] capped=[1, 37]
[span-overflow divisors] op=buf0 host_dim=1 required=3 legal=[1, 37] ... input_candidates=1
[span-overflow groups] op=buf1 started_new_pw_group split=[(0, 2/37, False)]
Unsupported: Cannot auto-tile buf0: it reads already auto-tiled producer(s) ['buf1'] ...
```

### What did you expect to happen?

The span-overflow tiler should tile a matmul together with its (auto-tiled) weight producer — a synchronized pointwise-producer → reduction-consumer group — so `F.linear` with a large vocab compiles. At minimum `V=87616` (a clean, in-cap, work-division-unsplittable case) should compile.

### How can we reproduce it?

```python
import torch, torch.nn.functional as F
x = torch.randn(1, 4096, dtype=torch.float16).to("spyre")
w = torch.randn(87616, 4096, dtype=torch.float16).to("spyre")   # 1369 = 37^2 sticks
out = torch.compile(lambda a, b: F.linear(a, b), dynamic=False)(x, w)
```
Run with `SPYRE_INDUCTOR_IGNORE_SPAN_OVERFLOW_HINTS=0 SPYRE_INDUCTOR_LOG=1 SPYRE_INDUCTOR_LOG_LEVEL=DEBUG`.

### Anything else?

- **Two separable gaps:** (a) producer/consumer (pointwise→reduction) group synchronization — blocks `49152`/`87616`; (b) prime / cap-exceeding stick counts — blocks `49216` (`769` needs 769 even tiles, and only stick-aligned split > `_MAX_AUTO_TILE_SPLIT_COUNT=64`). Gap (a) is the primary blocker for enabling the feature at all.
- Relevant code: `torch_spyre/_inductor/span_overflow_hint_analysis.py`, `torch_spyre/_inductor/coarse_tile.py` (`span_overflow_groups`, `_stamp_group`), `torch_spyre/_inductor/config.py` (`ignore_span_overflow_hints`).
- Related: #1918 (lm_head), #741 (pad dims for work division).

cc @Hemaprasannakc @dgrove-oss

**Comment by dgrove-oss** [comment]
I'm not sure we want to "scope creep" the auto tiling to go beyond single operations quite yet.  I think it would be functionally correct if the auto tiling produced two loops (one for the pointwise op and one for the matmul)?   The performance is non-ideal (HBM roundtrip for the intermediate value), but I don't see why it wouldn't work.   There's no scheduling problem since the first loop completes before the second one starts.

**Comment by mudhakar** [comment]
Status of the two gaps:

**Gap (a)** (matmul reading its auto-tiled weight producer): fixed by #3218 — the span-overflow tiler now lets the matmul join its producer's group (`87616`, `49152` compile + run; regressions green). The tiler is currently opt-in (`ignore_span_overflow_hints`); with gap (a) fixed I'd propose **enabling it by default, pending sign-off from @dgrove-oss and @Hemaprasannakc**.

**Gap (b)** (prime/large-prime-factor vocab stick count, e.g. `49216 → 769`): not fixable in torch-spyre **today** — work division / the span-overflow tiler split a stick dim only by *even divisors*, and a prime count has none usable. The in-compiler fix is **uneven/remainder tiling** (loop the stick dim in uneven chunks, no relayout) — future work. Until then, the fix is **host-side padding** to a smooth stick count, already done on the HF-adapters path by `pad_lm_head()` (`hf_adapters/hf_common.py`, e.g. `769 → 770`). Note: on-device *padding* is not an option — it re-incurs the same >256 MB EAR span on the pad op.

So HF-adapters + torch-spyre are covered: (a) via #3218, (b) via `pad_lm_head`. vLLM pathway tracked on #1918.
